addons/website_sale/controllers/main.py

```diff
@@ -246,28 +240,39 @@ def shop(self, page=0, category=None, search='', ppg=False, **post):
         if attrib_list:
             post['attrib'] = attrib_list
 
-        Product = request.env['product.template'].with_context(bin_size=True)
-
-        search_product = Product.search(domain)
+        domain = self._get_search_domain(search, attrib_values)
         website_domain = request.website.website_domain()
-        categs_domain = [('parent_id', '=', False)] + website_domain
-        if search:
-            search_categories = Category.search([('product_tmpl_ids', 'in', search_product.ids)] + website_domain).parents_and_self
-            categs_domain.append(('id', 'in', search_categories.ids))
-        else:
-            search_categories = Category
-        categs = Category.search(categs_domain)
+        categ_domain = expression.AND([domain, website_domain])
+        Product = request.env['product.template'].with_context(bin_size=True)
+        where_query = Product._where_calc(categ_domain)
+        Product._apply_ir_rules(where_query)
+        from_clause, where_clause, args = where_query.get_sql()
+        query = """
+            SELECT c.id, COUNT(DISTINCT product_template)
+            FROM %s
+            INNER JOIN %s ON product_template_id = product_template.id
+            INNER JOIN %s c ON product_public_category_id = c.id
+            WHERE %s
+            GROUP BY c.id
+        """ % (Category._fields['product_tmpl_ids'].relation, Product._table, Category._table, where_clause)
+        Category.flush()
+        request.env.cr.execute(query, args)
+        categs_nbr_prod = dict(request.env.cr.fetchall())
+        categs_ids = list(categs_nbr_prod.keys())
+        categs = Category.search([('id', 'parent_of', categs_ids), ('parent_id', '=', False)])
```

Comment:
If you don't get all categories here, there might be more queries on the recursive child print.

To be tested. And fix would be `with_prefetch(categs_ids)`. (make categs_ids a tuple above)

Reply:
Also since you have all categories in `categs_ids` you probably don't need `parent_of` and can just do `in` instead?

Reply:
I don't have categories with product(s) in there child only.